### PR TITLE
dev-libs/libei: New package 1.0.0, x11-wm/mutter, gnome-base/gnome-shell: Version bump to 45_beta1,  gnome-extra/gnome-shell-extensions: Version bump to 45_beta

### PR DESCRIPTION
**Pull request by flint2**
- dev-libs/libei: New package 1.0.0
- x11-wm/mutter: Version bump to 45_beta1
- gnome-base/gnome-shell: Version bump to 45_beta1
- gnome-extra/gnome-shell-extensions: Version bump to 45_beta

### CODE REVIEW

**Comment by mattst88** [comment]
For libei:

- Missing dependencies: `Program python3 (attr, jinja2) found: YES (/usr/bin/python3) modules: attr, jinja2`
- Useless sed: `sed "s@, '-Werror'@@" -i meson.build || die #744250`

Looking at this a bit more, this is just a sloppy copy-and-paste. Lots of wrong dependencies.

Please do better with new packages. This is not a good use of my time to point out each individual problem when you haven't really tried.

**Comment by flint2** [comment]
Just released mutter and gnome-shell 45.beta.1 
WIP, PR changed to Draft

**Comment by flint2** [comment]
libei test OK
```
ninja: Entering directory `/var/tmp/portage/dev-libs/libei-1.0.0/work/libei-1.0.0-build'
ninja: no work to do.
1/1 dtdcheck        OK              0.01s

Ok:                 1   
Expected Fail:      0   
Fail:               0   
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0   

```

This will go away on next release, is needed because of the exceptional naming:
[gnome-shell-45.beta**.1**.tar.xz](https://download.gnome.org/sources/gnome-shell/45/gnome-shell-45.beta.1.tar.xz)
[mutter-45.beta**.1**.tar.xz](https://download.gnome.org/sources/mutter/45/mutter-45.beta.1.tar.xz)
```
src_unpack() {
if [[ -n ${A} ]]; then
    unpack ${A}
    mv "${WORKDIR}"/gnome-shell-45.beta.1 "${WORKDIR}"/gnome-shell-45.beta
fi
}


``


mutter and gnome-shell test fails similar to previous versions
gnome-shell-extensions does not have tests

**Comment by mattst88** [comment]
How is libei building the tests for you? It appears to depend on https://github.com/nemequ/munit/ which isn't packaged in Gentoo. I think we will have to include a snapshot of it in libei because it hasn't seen a release since 2016, and it appears to be expected that libei will use it as a meson subproject.

**Comment by mattst88** [comment]
These are the changes I've made to libei locally so far:

```diff
diff --git a/dev-libs/libei/libei-1.0.0.ebuild b/dev-libs/libei/libei-1.0.0.ebuild
index e0329a9d9fdb..70193f4a44f8 100644
--- a/dev-libs/libei/libei-1.0.0.ebuild
+++ b/dev-libs/libei/libei-1.0.0.ebuild
@@ -13,25 +13,27 @@ SRC_URI="https://gitlab.freedesktop.org/libinput/${PN}/-/archive/${PV}/${P}.tar.
 LICENSE="MIT"
 SLOT="0/10"
 KEYWORDS="~alpha ~amd64 ~arm ~arm64 ~hppa ~ia64 ~loong ~m68k ~mips ~ppc ~ppc64 ~riscv ~s390 ~sparc x86"
-IUSE="doc test"
+IUSE="doc elogind systemd test"
 RESTRICT="!test? ( test )"
 
 RDEPEND="
 	>=dev-libs/libevdev-1.9.902
-	virtual/libudev:=
-	virtual/udev
-"
-DEPEND="
-	${RDEPEND}
-	test? ( >=dev-libs/check-0.9.10 )
+	|| (
+		systemd? ( >=sys-apps/systemd-237 )
+		elogind? ( >=sys-auth/elogind-237 )
+		sys-libs/basu
+	)
 "
+DEPEND="${RDEPEND}"
 BDEPEND="
 	virtual/pkgconfig
-	doc? (
+	test? (
 		$(python_gen_any_dep '
 			dev-python/attrs[${PYTHON_USEDEP}]
 			dev-python/jinja[${PYTHON_USEDEP}]
 		')
+	)
+	doc? (
 		>=app-doc/doxygen-1.8.3
 		>=media-gfx/graphviz-2.38.0
 	)
@@ -49,17 +51,24 @@ pkg_setup() {
 }
 
 src_configure() {
-	use doc && MYMESONARGS=-Ddocumentation=api
 	local emesonargs=(
+		-Ddocumentation=$(usex doc api "")
 		-Dliboeffis=enabled
 		$(meson_feature test tests)
 	)
+	if use systemd; then
+		emesonargs+=(-Dsd-bus-provider=libsystemd)
+	elif use elogind; then
+		emesonargs+=(-Dsd-bus-provider=libelogind)
+	else
+		emesonargs+=(-Dsd-bus-provider=basu)
+	fi
 	meson_src_configure
 }
 
 src_install() {
 	meson_src_install
-	if use doc ; then
+	if use doc; then
 		docinto html
 		dodoc -r "${BUILD_DIR}"/Documentation/.
 	fi
```

**Comment by mattst88** [comment]
I've filed https://gitlab.freedesktop.org/libinput/libei/-/issues/42

**Comment by mattst88** [comment]
You can drop

```
src_unpack() {
if [[ -n ${A} ]]; then
    unpack ${A}
    mv "${WORKDIR}"/gnome-shell-45.beta.1 "${WORKDIR}"/gnome-shell-45.beta
fi
}
```

by just giving the ebuilds the version `45_beta1`.

**Comment by flint2** [comment]
> > I've filed https://gitlab.freedesktop.org/libinput/libei/-/issues/42
> 
> ## Working now on this:
> 
> [Ian Douglas Scott](https://gitlab.freedesktop.org/ids1024) [@ids1024](https://gitlab.freedesktop.org/ids1024) · [58 minutes ago](https://gitlab.freedesktop.org/libinput/libei/-/issues/42#note_2052502) Contributor
> 
> For distro packaging, it's sufficient to add something like https://github.com/nemequ/munit/archive/fbbdf1467eb0d04a6ee465def2e529e4c87f2118.tar.gz as an additional source for the package, and extract that to subprojects/ (or link it).

Done in https://github.com/gentoo/gentoo/pull/32227/commits/4a8a75a96d5793516f79eef44d71affdcf07b7a0

**Comment by flint2** [comment]
Sorry about the rebase merging master, was an error updating the branch

**Comment by mattst88** [comment]
Are you actually testing this? I have to apply

```diff
diff --git a/dev-libs/libei/libei-1.0.0.ebuild b/dev-libs/libei/libei-1.0.0.ebuild
index 9bfa0c1374d8..be471a3bee6d 100644
--- a/dev-libs/libei/libei-1.0.0.ebuild
+++ b/dev-libs/libei/libei-1.0.0.ebuild
@@ -53,8 +53,8 @@ pkg_setup() {
 
 src_unpack() {
        if [[ -n ${A} ]]; then
-               unpack ${A}     
-               mv "${WORKDIR}"/munit-fbbdf1467eb0d04a6ee465def2e529e4c87f2118 "${WORKDIR}"/${P}/subprojects
+               unpack ${A}
+               mv "${WORKDIR}"/munit-fbbdf1467eb0d04a6ee465def2e529e4c87f2118 "${WORKDIR}"/${P}/subprojects/munit
                rm "${WORKDIR}"/${P}/subprojects/munit.wrap
        fi
 }
```

(ignore the change to the `unpack` line; it was fixing whitespace)

Without this, it fails exactly the same way as before -- unable to find `munit`.

With it, it makes it farther but then fails because it's missing 

```
C++ linker for the host machine: x86_64-pc-linux-gnu-g++ ld.bfd 2.40
Program python3 found: YES (/usr/bin/python3)
Program python3 (pytest, attr, structlog, dbusmock) found: NO modules: pytest, attr, dbusmock

test/meson.build:121:15: ERROR: python3 is missing modules: structlog
```

After adding that dependency, the tests finally build and run, but one fails because

```
FAILED test_scanner.py::TestScanner::test_cli_extra_data[yamlfile] - ModuleNotFoundError: No module named 'yaml'
```

With `dev-python/pyyaml` added as a test dependency all tests except `libei:python / protocol-test-valgrind` pass (and that test times out).

File: gnome-base/gnome-shell/gnome-shell-45_beta1.ebuild [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+# libXfixes-5.0 needed for pointer barriers and #include <X11/extensions/Xfixes.h>
+DEPEND="
+	>=gnome-extra/evolution-data-server-3.46.0:=
+	>=app-crypt/gcr-3.90.0:4=[introspection]
+	>=dev-libs/glib-2.68:2
+	>=dev-libs/gobject-introspection-1.49.1:=
+	>=dev-libs/gjs-1.73.1[cairo]
+	>=gui-libs/gtk-4:4[introspection]
+	>=x11-wm/mutter-45_beta:0/13[introspection,test?]
```
Comment (mattst88) [inline gnome-base/gnome-shell/gnome-shell-45_beta1.ebuild:29]: `s/beta/beta1/`

File: dev-libs/libei/libei-1.0.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+PYTHON_COMPAT=( python3_{10..12} )
+inherit meson optfeature python-any-r1 udev
+
+DESCRIPTION="Library for Emulated Input, primarily aimed at the Wayland stack"
+HOMEPAGE="https://gitlab.freedesktop.org/libinput/libei"
+SRC_URI="https://gitlab.freedesktop.org/libinput/${PN}/-/archive/${PV}/${P}.tar.bz2"
+SRC_URI+=" https://github.com/nemequ/munit/archive/fbbdf1467eb0d04a6ee465def2e529e4c87f2118.tar.gz -> munit-9999.tar.gz"
+
+LICENSE="MIT"
+SLOT="0/10"
```
Comment (mattst88) [inline dev-libs/libei/libei-1.0.0.ebuild:15]: I don't see anything in the package that indicates `10` is the correct subslot. I don't even know that we need a subslot.

File: dev-libs/libei/libei-1.0.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,92 @@
+# Copyright 2014-2023 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+PYTHON_COMPAT=( python3_{10..12} )
+inherit meson optfeature python-any-r1 udev
```
Comment (mattst88) [inline dev-libs/libei/libei-1.0.0.ebuild:7]: No need for `optfeature` or `udev`.

File: dev-libs/libei/libei-1.0.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	fi
+}
+
+pkg_postinst() {
+	udev_reload
+}
+
+pkg_postrm() {
+	udev_reload
+}
```
Comment (mattst88) [inline dev-libs/libei/libei-1.0.0.ebuild:92]: No need for either of these functions.

File: gnome-base/gnome-shell/gnome-shell-45_beta1.ebuild [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+PYTHON_COMPAT=( python3_{10..12} )
+
+inherit gnome.org gnome2-utils meson optfeature python-single-r1 virtualx xdg
+
+DESCRIPTION="Provides core UI functions for the GNOME desktop"
+HOMEPAGE="https://wiki.gnome.org/Projects/GnomeShell https://gitlab.gnome.org/GNOME/gnome-shell"
+
+LICENSE="GPL-2+ LGPL-2+"
+SLOT="0"
+IUSE="cameramonitor elogind gtk-doc +ibus +networkmanager systemd test"
```
Comment (mattst88) [inline gnome-base/gnome-shell/gnome-shell-45_beta1.ebuild:14]: You're adding `IUSE=cameramonitor` but making the dependency on pipewire unconditional.

I'd suggest just naming the USE flag "pipewire" and making the pipewire dep conditional on that.

File: x11-wm/mutter/mutter-45_beta1.ebuild [hunk 6]
```diff
@@ -0,0 +1,235 @@
+# Copyright 1999-2023 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+PYTHON_COMPAT=( python3_{9..11} )
```
Comment (mattst88) [inline x11-wm/mutter/mutter-45_beta1.ebuild:5]: I don't know if there's something blocking add Py3.12, but we should remove 9 regardless.

**Comment by flint2** [comment]
> ```diff
> jects/mu
> ```

I'm probably testing it wrong
ebuild libei-1.0.0.ebuild prepare
ebuild libei-1.0.0.ebuild test

```
>>> Test phase: dev-libs/libei-1.0.0
meson test --print-errorlogs -C /var/tmp/portage/dev-libs/libei-1.0.0/work/libei-1.0.0-build --num-processes 12
ninja: Entering directory `/var/tmp/portage/dev-libs/libei-1.0.0/work/libei-1.0.0-build'
ninja: no work to do.
1/1 dtdcheck        OK              0.01s

Ok:                 1   
Expected Fail:      0   
Fail:               0   
Unexpected Pass:    0   
Skipped:            0   
Timeout:            0   
```

I´ve missed 
ebuild libei-1.0.0.ebuild clean
before test

**Comment by flint2** [comment]
> Are you actually testing this? I have to apply
> 
> ```diff
> diff --git a/dev-libs/libei/libei-1.0.0.ebuild b/dev-libs/libei/libei-1.0.0.ebuild
> index 9bfa0c1374d8..be471a3bee6d 100644
> --- a/dev-libs/libei/libei-1.0.0.ebuild
> +++ b/dev-libs/libei/libei-1.0.0.ebuild
> @@ -53,8 +53,8 @@ pkg_setup() {
>  
>  src_unpack() {
>         if [[ -n ${A} ]]; then
> -               unpack ${A}     
> -               mv "${WORKDIR}"/munit-fbbdf1467eb0d04a6ee465def2e529e4c87f2118 "${WORKDIR}"/${P}/subprojects
> +               unpack ${A}
> +               mv "${WORKDIR}"/munit-fbbdf1467eb0d04a6ee465def2e529e4c87f2118 "${WORKDIR}"/${P}/subprojects/munit
>                 rm "${WORKDIR}"/${P}/subprojects/munit.wrap
>         fi
>  }
> ```
> 
> (ignore the change to the `unpack` line; it was fixing whitespace)
> 
> Without this, it fails exactly the same way as before -- unable to find `munit`.
> 
> With it, it makes it farther but then fails because it's missing
> 
> ```
> C++ linker for the host machine: x86_64-pc-linux-gnu-g++ ld.bfd 2.40
> Program python3 found: YES (/usr/bin/python3)
> Program python3 (pytest, attr, structlog, dbusmock) found: NO modules: pytest, attr, dbusmock
> 
> test/meson.build:121:15: ERROR: python3 is missing modules: structlog
> ```
> 
> After adding that dependency, the tests finally build and run, but one fails because
> 
> ```
> FAILED test_scanner.py::TestScanner::test_cli_extra_data[yamlfile] - ModuleNotFoundError: No module named 'yaml'
> ```
> 
> With `dev-python/pyyaml` added as a test dependency all tests except `libei:python / protocol-test-valgrind` pass (and that test times out).

Now I have the same behavior, even bare clone an test is giving timeout

```
git clone https://gitlab.freedesktop.org/libinput/libei.git
meson libei /tmp/libei
ninja -C /tmp/libei test        
ninja: Entering directory `/tmp/libei'
[0/1] Running all tests.
 1/10 libei / unit-tests-utils                      OK              0.02s
 2/10 libei / unit-tests-ei                         OK              0.02s
 3/10 libei / unit-tests-eis                        OK              0.02s
 4/10 libei / unit-tests-oeffis                     OK              0.02s
 5/10 libei / dtdcheck                              OK              0.02s
 6/10 libei / eierpecken                            OK              0.04s
 7/10 libei:python / scanner-pytest                 OK              0.31s
 8/10 libei:python / oeffis-pytest                  OK              2.45s
 9/10 libei:python / protocol-test                  OK              6.57s
libei:python / protocol-test-valgrind time out (After 30 seconds)
10/10 libei:python / protocol-test-valgrind         TIMEOUT        30.01s   killed by signal 15 SIGTERM
>>> LIBEI_USE_VALGRIND=1 MALLOC_PERTURB_=142 /usr/bin/pytest --verbose --log-level=DEBUG -k TestEiProtocol
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
============================= test session starts ==============================
platform linux -- Python 3.11.4, pytest-7.4.0, pluggy-1.2.0 -- /usr/bin/python3.11
cachedir: .pytest_cache
rootdir: /tmp/libei
plugins: pkgcore-0.12.22
collecting ... collected 37 items / 8 deselected / 29 selected

test/test_protocol.py::TestEiProtocol::test_server_sends_version_event_immediately 
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――


Summary of Failures:

10/10 libei:python / protocol-test-valgrind TIMEOUT        30.01s   killed by signal 15 SIGTERM

Ok:                 9   
Expected Fail:      0   
Fail:               0   
Unexpected Pass:    0   
Skipped:            0   
Timeout:            1   

Full log written to /tmp/libei/meson-logs/testlog.txt
FAILED: meson-internal__test 
/usr/lib/python-exec/python3.11/meson test --no-rebuild --print-errorlogs
ninja: build stopped: subcommand failed.
```

**Comment by mattst88** [comment]
Oh, I see. Yeah, you have to issue the `test` command to `ebuild` initially. If you `unpack`/`prepare`/`configure`/`compile` without `test` it will not subsequently run tests.

File: dev-libs/libei/libei-1.0.0.ebuild [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+		emesonargs+=(-Dsd-bus-provider=basu)
+	fi
+	meson_src_configure
+}
+
+src_install() {
+	meson_src_install
+	if use doc; then
+		docinto html
+		dodoc -r "${BUILD_DIR}"/Documentation/.
```
Comment (mattst88) [inline dev-libs/libei/libei-1.0.0.ebuild:88]: This fails...

Looking at `doc/api/meson.build`:

```
custom_target('doxygen',
	      input : [ doxyfile, mainpage ] + src_doxygen + doxyfiles,
	      output : [ 'doc' ],
	      command : [ doxygen, doxyfile ],
	      install : false,
	      depends: [ mainpage ],
	      build_by_default : true)
```

The docs aren't installed even if they're generated. So I'd just pass `-Ddocumentation=""` and remove `IUSE=doc` (and drop `src_install`).

**Pull request closed without merging**